califa/calibration/R3BCalifaCrystalCal2Cluster.cxx

```diff
@@ -138,193 +242,317 @@ InitStatus R3BCalifaCrystalCal2Cluster::ReInit()
+
+    vector<Int_t> usedCrystals;
+
+    Double_t cryEnergy;
+    Int_t cryId;
+
+    for (Int_t i = 0; i < numCrystalHits; i++)
     {
-        case RECT:
-        { // rectangular window
-            if (TMath::Abs(vref.Theta() - vhit.Theta()) < fDeltaPolar &&
-                circleAbs(vref.Phi() - vhit.Phi()) < fDeltaAzimuthal)
+
+        cryId = dynamic_cast<R3BCalifaCrystalCalData*>(fCrystalCalData->At(i))->GetCrystalId();
+        cryEnergy = dynamic_cast<R3BCalifaCrystalCalData*>(fCrystalCalData->At(i))->GetEnergy();
+
+        if (cryEnergy >= fCrystalThreshold)
+            allCrystalVec.push_back(dynamic_cast<R3BCalifaCrystalCalData*>(fCrystalCalData->At(i)));
+
+        /* -------- Real data ------- */
+        if (!fSimulation)
+        {
+
+            if (cryEnergy >= fGammaClusterThreshold && cryEnergy <= fSaturation)
+                gammaCandidatesVec.push_back(dynamic_cast<R3BCalifaCrystalCalData*>(fCrystalCalData->At(i)));
+
+            if (cryId > fTotalCrystals && cryEnergy >= fProtonClusterThreshold)
+                protonCandidatesVec.push_back(dynamic_cast<R3BCalifaCrystalCalData*>(fCrystalCalData->At(i)));
+
+            if (cryId <= fTotalCrystals && cryEnergy > fSaturation)
+                saturatedCandidatesVec.push_back(dynamic_cast<R3BCalifaCrystalCalData*>(fCrystalCalData->At(i)));
+        }
+
+        /* -------- Simulation data ------- */
+        else
+        {
+
+            if (cryEnergy >= fGammaClusterThreshold && cryEnergy < fProtonClusterThreshold)
+                gammaCandidatesVec.push_back(dynamic_cast<R3BCalifaCrystalCalData*>(fCrystalCalData->At(i)));
+
+            if (cryEnergy >= fProtonClusterThreshold)
+                protonCandidatesVec.push_back(dynamic_cast<R3BCalifaCrystalCalData*>(fCrystalCalData->At(i)));
+        }
+    }
+
+    /* ----- Remove duplicate entries in gamma clusters ----- */
+    for (int k = 0; k < gammaCandidatesVec.size(); k++)
+    {
+        for (int s = 0; s < gammaCandidatesVec.size(); s++)
+        {
+
+            if (gammaCandidatesVec.at(k)->GetCrystalId() == (gammaCandidatesVec.at(s)->GetCrystalId() - fTotalCrystals))
+                gammaCandidatesVec.erase(gammaCandidatesVec.begin() + s);
+        }
+    }
+
+    // Sort all vectors by energy
+    std::sort(gammaCandidatesVec.begin(), gammaCandidatesVec.end(), compareByEnergy);
+    std::sort(protonCandidatesVec.begin(), protonCandidatesVec.end(), compareByEnergy);
+
+    TVector3 mother_angles, angles;
+    Double_t fRandTheta, fRandPhi;
+
+    while (protonCandidatesVec.size())
+    {
+
+        Int_t motherId = protonCandidatesVec.at(0)->GetCrystalId();
+
+        califa_candidate cluster = { motherId, vector<Int_t>(), 0.0, 0.0, 0.0, 0.0, 0.0 };
+
+        if (motherId <= fTotalCrystals)
+            mother_angles = fCalifaGeo->GetAngles(motherId) - fCalifatoTargetPos;
+
+        else
+            mother_angles = fCalifaGeo->GetAngles(motherId - fTotalCrystals) - fCalifatoTargetPos;
```

Comment:
I think this is dangerous to make if statements all the time. Why not define an integer as
`Int_t motherCrysId = motherId<=fTotalCrystals?motherId:motherId-fTotalCrysltals;`?

Reply:
I would prefer 
`` TVector3 mothervec=fCalifaGeo->GetVector(mother.GetLowGainCrystalId()) -fCalifaToTargetPos;``

Reply:
I believe in the evilness of ternary operators. Not on my watch. Why an if is dangerous?

Reply:
The reason I thought it is dangerous is that you have to deduce the crystal ID every time you need it. I am not saying we should use ternary operators but I think it's better to define a variable representing ID first, and then use this within the scope. But, as @klenze  mentions, if you can prepare the CrystalCal class to handle both proton and gamma ranges within one class, that would be even more straightforward.